opal/mca/common/cuda/common_cuda.c

```diff
@@ -1730,6 +1741,46 @@ static float mydifftime(opal_timer_t ts_start, opal_timer_t ts_end) {
 }
 #endif /* OPAL_ENABLE_DEBUG */
 
+static CUmemorytype mca_common_cuda_check_host_numa(CUdeviceptr dbuf)
+{
+#if OPAL_CUDA_HOST_NUMA_SUPPORT
+    CUmemAllocationProp prop;
+    CUmemLocation location;
+    CUresult result;
+    unsigned long long flags;
+    CUmemGenericAllocationHandle alloc_handle;
+    /* Check if memory is allocated using VMM API and see if host memory needs
+     * to be treated as pinned device memory */
+    result = cuFunc.cuMemRetainAllocationHandle(&alloc_handle, (void*)dbuf);
```

Comment:
This looks not only overly complicated but also incorrect. 

Regarding correctness: according to the CUDA documentation each call the `cuMemRetainAllocationHandle` must be matched with a call to `cuMemRelease`, which i don't see in this PR. This will result in the memory region referenced here not being able to be released.

What exactly do you get from the combination `cuMemRetainAllocationHandle` + `cuMemGetAllocationPropertiesFromHandle` that you could not have obtained from `cuMemGetAccess` ?